### DISCUSSION TITLE
Is it appropriate to mix Polly and FluentResults?

### DISCUSSION THREAD

**Discussion by dguisinger** [App-vNext/Polly#2159] [Q&A] [answered]
I'm looking at evaluating Polly.  Since it performs error some handling to do things such as retries, would it work well in a project where a lot of error handling/propagation is already performed by FluentResults?  Or would that be a bad combination and not work as intended?

**Comment by peter-csala** [comment]
Hi Dan, I have used Polly and FluentResults separately (in different projects). I don't see any fundamental problem to use both together. `Result.Try` might be tricky in some cases but others should work flawlessly IMHO. 

If you have a particular problem that you try to solve with Polly and FluentResults then please share with us so, we can create a POC.

**Reply by dguisinger** [reply]
I'll give it a try, I think I was mainly concerned about whether or not anything would work if I used Result.Try to catch exceptions in code that is being retried.

**Reply by peter-csala** [reply, marked as answer]
[Here](https://dotnetfiddle.net/ciODWo) I have put together an example:
```cs
public class Program
{
	public static void Main()
	{
		var retry = new ResiliencePipelineBuilder<Result<int>>()
			.AddRetry(new ()
			{
				ShouldHandle = static args => args.Outcome switch
				{
						{ Result: { IsFailed: true, Errors: List<IError> errs } } when (errs.OfType<ExceptionalError>().Any()) => PredicateResult.True(),
						_ => PredicateResult.False()
				},
				MaxRetryAttempts = 5,
				OnRetry = static args =>
				{
					 Console.WriteLine("OnRetry, Attempt: {0}", args.AttemptNumber);
					return default;
				}
			})
			.Build();
		
		var res = retry.Execute(Get);
		res.Dump();
	}
	
	private static Result<int> Get()
	{
		return Result.Try<int>(GetCore, ex => 
			Random.Shared.Next() % 2 == 0 
				? new ExceptionalError("Something went wrong", ex)
				: new CustomError("Something went terribly wrong")	
			);
	}
	
	private static int GetCore()
	{
		if(Random.Shared.Next() % 2 == 0)
		{
			throw new Exception("Not worked");
		}
		return 42;
	}	
}
```
```cs
public class CustomError : Error
{
    public CustomError(string message)
        : base(message)
    { 
    }
}
```
- `GetCore` either returns the meaning of life or throws an exception
- `Get` wraps the outcome of the `GetCore` into a `Result<int>` object
  - If `GetCore` returned 42 then a successful `Result<int>` is created
  - If `GetCore` threw an exception then either an `ExceptionalError` or a `CustomError` is created
- `Main` defines a retry strategy which should trigger 
   - only if the `Result` is failed and contains an  `ExceptionalError`

I know, it is a really made up example but it shows that Polly and FluentResults can work together smoothly.

**Reply by dguisinger** [reply]
Beautiful, thank you for putting that together.  With no experience yet with Polly I wasn't sure if I had to let exceptions flow down.  Now I see I don't, that helped a lot.
